sdk/core/core/src/az_aad_private.h

```diff
@@ -38,6 +38,22 @@ AZ_NODISCARD bool _az_token_expired(_az_token const* token);
 
 AZ_NODISCARD az_result _az_token_set(_az_token* self, _az_token const* new_token);
 
+/**
+ * @brief Copies a URL in the \p source #az_span to the \p destination #az_span by url-encoding the
+ * \p source span characters.
+ *
+ * @param[in] destination The #az_span whose bytes will receive the url-encoded source.
+ * @param[in] source The #az_span containing the non-url-encoded bytes.
+ * @param[out] out_span A pointer to an #az_span that receives the remainder of the \p destination
+ * #az_span after the url-encoded \p source has been copied.
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if successful
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the \p destination is not big enough to contain
+ * the encoded bytes
+ */
+AZ_NODISCARD az_result
+_az_span_copy_url_encode(az_span destination, az_span source, az_span* out_span);
```

Comment:
I believe the initial intent of the issue was to move this into a static within az_aad and not expose it from any headers. However, given we have extensive tests for this (because the logic isn't trivial), I talked with @JeffreyRichter about this during the span redesign (https://github.com/Azure/azure-sdk-for-c/pull/583) and decided to move it from the public span header to a private span header. I think that is good enough.

Moving, what is essentially a span-based helper API, from span_private to aad_private isn't that useful, imo. What do do you think @antkmsft ?

Reply:
I would keep it close to the only functionality that uses it. It is not public API yet, so we can keep it here. If in the future there is another place using it, we should move it to span_private. Currently, if you remove AAD, you remove the only user of the url_encode, together with the function which is good. Plus hen we have it in the same file, I can more freely remove precondition checks, which I did.

Reply:
>  Currently, if you remove AAD, you remove the only user of the url_encode, together with the function which is good.

That can be said about a lot of the private/internal functions. If the component/components that use it are gone, we'd want to remove those too, but:
1) How likely is that to happen (where we remove AAD)
2) Wouldn't we want to go through the exercise of removing unreachable code in private/internals anyway for APIs that are not trivially detected as "single use".

I don't know if that is a good reason to move it. Where does this API naturally belong? If you think that it is in aad, then it makes sense. However, what happens if we get another usage of url encoding outside of aad? It doesn't make sense to include aad_private to get this API.

> Plus hen we have it in the same file, I can more freely remove precondition checks, which I did.

I am not a fan of that and might be another reason to avoid this move. This tends to go against single responsibility (and having reusable, testable, modular components).

Reply:
Maybe I should change the prefix from `_span` to solething like `_url_encode`. This function is not really about spans IMO. Yes it takes spans as parameters, but it is not a function like slice() or append_character().

Reply:
Sounds good.

Reply:
Check out the new naming, looks good? :)